### PR TITLE
check-for-changes wait for settle

### PR DESCRIPTION
**Pull request by NorseGaud**
Original PR this was split from: https://github.com/docker-mailserver/docker-mailserver/pull/2157

Goal: Have check-for-changes.sh wait for changes to "settle" (checksum comparison has to not have changed over two different checks with 5-second gaps between) before modifying and restarting anything.

Why: The eventual API (https://github.com/docker-mailserver/docker-mailserver-admin/issues/3) will allow users to make frequent changes to the config and we don't want them triggering service restarts until those requests/changes settle down.

### CODE REVIEW

**Comment by polarathene** [comment]
I haven't looked over this PR update, but chiming in to say I still feel strongly about [what I said in the previous PR](https://github.com/docker-mailserver/docker-mailserver/pull/2157#pullrequestreview-817286005) back in Nov 2021. I've mentioned it a few times but had no feedback.

I was hoping to have started working on that by now, but lacked resources. That should be resolved sometime this month thankfully, and I can start contributing again.

Did you want to discuss the alternate route I think we should pursue, or for the time being get your approach in first?

**Comment by NorseGaud** [comment]
> I haven't looked over this PR update, but chiming in to say I still feel strongly about [what I said in the previous PR](https://github.com/docker-mailserver/docker-mailserver/pull/2157#pullrequestreview-817286005) back in Nov 2021. I've mentioned it a few times but had no feedback.
> 
> I was hoping to have started working on that by now, but lacked resources. That should be resolved sometime this month thankfully, and I can start contributing again.
> 
> Did you want to discuss the alternate route I think we should pursue, or for the time being get your approach in first?

Do you mean `watchexec`? It adds a dependency which we'd have to wait on fixes for should something break or stop working. I don't think it's a good idea. This isn't that hard to solve in just pure bash.

**Comment by casperklein** [comment]
If I understand correctly, we would also need a work-around for at least NFS. I think we should aim a one-for-all solution (if possible).

**Comment by NorseGaud** [comment]
> If I understand correctly, we would also need a work-around for at least NFS. I think we should aim a one-for-all solution (if possible).

What is wrong with NFS? I'd like to keep this PR to only the essentials if possible. These massive PRs (which I'm responsible for) are painful for us all.

**Comment by casperklein** [comment]
I've updated my post above, I hope this makes it clearer 😉

**Comment by polarathene** [comment]
> Regarding watchexec: If I understand correctly, we would also need a work-around for at least NFS. I think we should aim a one-for-all solution (if possible).

`watchexec` already handles that and switches to polling instead of more efficient file watcher.

It additionally has a debounce setting, so that it won't run any script if an event triggers again within a duration (must be idle for X time after last event to run script), and a throttle IIRC to ensure that it will not debounce forever, at least once per Y time.

---

> Do you mean `watchexec`? It adds a dependency which we'd have to wait on fixes for should something break or stop working.

Yes, `watchexec`. What fixes do you expect to wait on or breakage to happen? It's a rust binary, I don't see that causing any major concerns.

The developer is quite friendly / responsive, chances are if a problem needed to be investigated/addressed, it'd be done sooner on their end than us with bash given how long some bug-fixes/features can take for us volunteer maintainers to complete.

Regardless, if we take your approach first for now, you'd have a fallback option to rollback to and maintain, but I don't think that'd be necessary in practice.

---

> I don't think it's a good idea. This isn't that hard to solve in just pure bash.

It might not be that hard to add to the existing bash support, but overall we could reduce a fair amount of bash to maintain, which I consider a decent reason in this case since we've already had bash introduce breakage that wasn't easy to identify and resolve quickly..

That was due to a recent base image update caused memory leak because `check-for-changes` was running the `sha512sum` via subshell, and this was possibly related to an incompatibility with `libc` update with older kernels (_from about a year ago or so, but still common for servers_).

We've already had other maintainers (myself included) find this area a bit tricky/confusing in the past to understand (_thankfully a fair amount of refactoring by you and me has greatly improved that at least_). I don't know about you, but my memory on such inner workings / layers in a codebase can get foggy over time, and I know I won't be as active of a maintainer  after this year, so someone else needs to retain/accumulate that knowledge and so forth.

As time goes by digging through git blame (_especially with big changes and logic being split into separate files or renamed/relocated_), trying to learn from past contributors/PRs can be frustrating, I did that extensively for this area especially for file locking logic.

IMO, simplifying this sort of thing as much as possible so that it's easy to grok and maintain is more important than it being handled in pure bash. Similar to the python `acme_extract`, just because it could be done in bash doesn't mean it's appropriate to. Bash makes sense elsewhere in the project, @georglauterbach and @casperklein do a stellar job keeping that in tip-top shape.

This PR is probably fine for now, but my opinion is it only further complicates the logic / control-flow. If `watchexec` reduces maintenance cost, provides better polling approach (_compared to managing loops+sleeps_) as a fallback all the while letting us leverage more effective file watchers for the majority of users... I'm quite down for that :sweat_smile: 

Not sure if you're aware but related tests updated by myself and @georglauterbach and perhaps your own experience have had failures due to the sleep approach required by current `check-for-changes` approach, and increasing the sleep or bytes checked by logs to accommodate has accidentally resulted in false positives where the author didn't consider the logs they were checking might be unrelated older logs from a previous change event... and probably introduced some race conditions.

---

## TL;DR

`watchexec` has the promise of:

- Less maintenance woes.
- Probably faster resolution to any issues compared to existing history with the current bash script.
- File watching with automatic polling fallback for users that need it (eg NFS).

Bash solution may be simple, but that doesn't always make it appropriate to prefer bash for the sake of it. Personally, `watchexec` seems it'd be simpler and more flexible to maintain on our end. I'm more concerned about `check-for-changes` logic as a whole, rather than what this PR alone is solving.

I'm not against reviewing and merging this PR in the meantime though, but I think there's a strong case for `watchexec` at a later date. We can always roll back to your bash approach if I'm wrong about `watchexec`.

**Review by polarathene — CHANGES_REQUESTED** [review]
LGTM :+1: Just a few minor changes to consider.

File: target/scripts/check-for-changes.sh [hunk 1]
```diff
@@ -48,108 +48,128 @@ sleep 10
...
-    CHANGED=$(grep -Fxvf "${CHKSUM_FILE}" "${CHKSUM_FILE}.new" | sed 's/^[^ ]\+  //')
+    # If the current run's CMP_RESULT is equal to the LAST_CMP_RESULT, changes have settled and we can safely proceed
+    # This prevents multiple restarts of postfix from happening all at once and causing problems or delays
+    if [[ "${CMP_RESULT}" == "${LAST_CMP_RESULT}" ]]
+    then
 
-    # TODO Perform updates below conditionally too
-    # Also note that changes are performed in place and are not atomic
-    # We should fix that and write to temporary files, stop, swap and start
+      _notify 'inf' "${LOG_DATE} Changes settled... Applying changes and restarting services..."
```
[the remark this answers is not in the document]
Comment (NorseGaud) [reply target/scripts/check-for-changes.sh:68]: Bad copy and paste. Fixed

File: test/mail_changedetector.bats [hunk 2]
```diff
@@ -52,12 +52,16 @@ function teardown_file() {
...
+  assert_output --partial 'Changes settled'
   assert_output --partial "Removed lock"
   run docker exec mail_changedetector_two /bin/bash -c "supervisorctl tail -3000 changedetector"
   assert_output --partial "postfix: stopped"
   assert_output --partial "postfix: started"
-  assert_output --partial "Change detected"
+  assert_output --partial 'Changes detected'
+  sleep 10
+  assert_output --partial 'Changes settled'
   assert_output --partial "Removed lock"
```
[the remark this answers is not in the document]
Comment (NorseGaud) [reply test/mail_changedetector.bats:65]: I have to change a bad habit of changing more than is necessary (essentialism; I highly recommend the book) and won't be doing things that aren't necessary to the goal anymore. I'm more than happy to approve a PR of yours if you want this done. But keep in mind, the rest of the tests in this file don't put blank lines between them.

**Comment by NorseGaud** [comment]
Seeing some flakiness in tests. I'm going to run them a few more times to see what and how often it fails.

**Comment by casperklein** [comment]
As always, good summary @polarathene 👍 

>>Regarding watchexec: If I understand correctly, we would also need a work-around for at least NFS. I think we should aim a one-for-all solution (if possible).

>watchexec already handles that and switches to polling instead of more efficient file watcher. It additionally has a debounce setting, so that it won't run any script if an event triggers again within a duration (must be idle for X time after last event to run script), and a throttle IIRC to ensure that it will not debounce forever, at least once per Y time.

This sounds like a jack of all trades 😄 If that works reliable and the code base can be considerably reduced, I think it's worth to evaluate this solution.

But until then, there are no reasons to not improve the current script with this PR.

**Review by polarathene — DISMISSED** [review]

**Comment by NorseGaud** [comment]
Flakey test: 

```
not ok 183 checking relay hosts: default mapping is added from env vars for new virtual user entry in 11546ms
[191](https://github.com/docker-mailserver/docker-mailserver/runs/5399413386?check_suite_focus=true#step:8:191)
# (from function `run_until_success_or_timeout' in file test/test_helper/common.bash, line 58,
[192](https://github.com/docker-mailserver/docker-mailserver/runs/5399413386?check_suite_focus=true#step:8:192)
#  in test file test/mail_with_relays.bats, line 62)
[193](https://github.com/docker-mailserver/docker-mailserver/runs/5399413386?check_suite_focus=true#step:8:193)
#   `run_until_success_or_timeout 10 docker exec mail_with_relays grep -e domain2.tld /etc/postfix/relayhost_map' failed
```

**Comment by georglauterbach** [comment]
> Flakey test: 
> 
> 
> 
> ```
> 
> not ok 183 checking relay hosts: default mapping is added from env vars for new virtual user entry in 11546ms
> 
> [191](https://github.com/docker-mailserver/docker-mailserver/runs/5399413386?check_suite_focus=true#step:8:191)
> 
> # (from function `run_until_success_or_timeout' in file test/test_helper/common.bash, line 58,
> 
> [192](https://github.com/docker-mailserver/docker-mailserver/runs/5399413386?check_suite_focus=true#step:8:192)
> 
> #  in test file test/mail_with_relays.bats, line 62)
> 
> [193](https://github.com/docker-mailserver/docker-mailserver/runs/5399413386?check_suite_focus=true#step:8:193)
> 
> #   `run_until_success_or_timeout 10 docker exec mail_with_relays grep -e domain2.tld /etc/postfix/relayhost_map' failed
> 
> ```

Difficult to say but I have never experienced this test to be flakey with other PRs or the current code base. How certain we be that this test really is flakey and it's not the PR's fault?

**Comment by polarathene** [comment]
From the shared log, it's possible that the timeout of 10 seconds is too short due to the extra sleeps and slower CI hardware vs local testing (I recall this being a gotcha in the past for other maintainers). If the test is consistently passing locally it's likely the cause is CI environment.

There was many more failures in the failed commit earlier, likely all due to this reason.. It may be a good indication to take caution with eventual merging if it's going to ripple through to other PRs?

For tests that have too short timeouts, we could increase them (presumably safe), but others may be failing due to assumptions that delays between service availability wouldn't be as long without any explicit timing? :man_shrugging:

**Comment by casperklein** [comment]
> How certain we be that this test really is flakey and it's not the PR's fault?

We merged a lot PRs in the last days and this test was never flaky IIRC.

**Comment by NorseGaud** [comment]
> > Flakey test: 
> > 
> > 
> > 
> > ```
> > 
> > not ok 183 checking relay hosts: default mapping is added from env vars for new virtual user entry in 11546ms
> > 
> > [191](https://github.com/docker-mailserver/docker-mailserver/runs/5399413386?check_suite_focus=true#step:8:191)
> > 
> > # (from function `run_until_success_or_timeout' in file test/test_helper/common.bash, line 58,
> > 
> > [192](https://github.com/docker-mailserver/docker-mailserver/runs/5399413386?check_suite_focus=true#step:8:192)
> > 
> > #  in test file test/mail_with_relays.bats, line 62)
> > 
> > [193](https://github.com/docker-mailserver/docker-mailserver/runs/5399413386?check_suite_focus=true#step:8:193)
> > 
> > #   `run_until_success_or_timeout 10 docker exec mail_with_relays grep -e domain2.tld /etc/postfix/relayhost_map' failed
> > 
> > ```
> 
> Difficult to say but I have never experienced this test to be flakey with other PRs or the current code base. How certain we be that this test really is flakey and it's not the PR's fault?

When I say flakey, I mean likely due to this PRs changes

**Comment by NorseGaud** [comment]
> > How certain we be that this test really is flakey and it's not the PR's fault?
> 
> 
> 
> We merged a lot PRs in the last days and this test was never flaky IIRC.
> 
> 
> 
> PS: This test also fails (reliable) when I run the tests locally.

Running this PR?

**Comment by casperklein** [comment]
Yes, with this commit [94a72dd](https://github.com/docker-mailserver/docker-mailserver/pull/2443/commits/94a72dd52cb43a8d89ca98fc7aff49c832348714)

**Comment by NorseGaud** [comment]
> Yes, with this commit [94a72dd](https://github.com/docker-mailserver/docker-mailserver/pull/2443/commits/94a72dd52cb43a8d89ca98fc7aff49c832348714)

Thanks, give me a bit to dig into this.

**Comment by NorseGaud** [comment]
To everyone watching this PR: I figured that I would change the shebang in setup.sh as part of this PR since it was blocking me running tests on macOS (https://github.com/docker-mailserver/docker-mailserver/pull/2443/commits/3964c981345af66474c93b919e39bde6cb4e5c92). It is the same issue described in https://github.com/docker-mailserver/docker-mailserver/pull/2448. I don't expect that it would be a breaking change... What do you think?

File: test/mail_with_relays.bats [hunk 3]
```diff
@@ -47,7 +47,7 @@ function teardown_file() {
   run docker exec mail_with_relays grep -e domain2.tld /etc/postfix/relayhost_map
   assert_output ''
   run ./setup.sh -c mail_with_relays alias add user2@domain2.tld user2@domaintwo.tld
-  run_until_success_or_timeout 10 docker exec mail_with_relays grep -e domain2.tld /etc/postfix/relayhost_map
+  run_until_success_or_timeout 11 docker exec mail_with_relays grep -e domain2.tld /etc/postfix/relayhost_map
```
Comment (NorseGaud) [inline test/mail_with_relays.bats:50]: Turns out adding aliases takes 1 second longer than anything else :)

**Comment by NorseGaud** [comment]
(Just making a note of this; no action necessary)
Found a failing test on macOS:

```bash
 ✗ checking ssl: manual cert changes are picked up by check-for-changes
   (in test file test/mail_ssl_manual.bats, line 112)
     `sed -i '/someThingsChangedHere/d' "$(pwd)/test/test-files/ssl/${DOMAIN_SSL_MANUAL}/with_ca/ecdsa/key.ecdsa.pem"' failed
   [[ TASKS ]]  2022-03-04 13:47:18 Start check-for-changes script.
   [[  INF  ]]  2022-03-04 13:47:18 Using postmaster address postmaster@example.test
   [[  INF  ]]  2022-03-04 13:47:28 check-for-changes is ready
   [[  INF  ]]  2022-03-04 13:47:28 Changes detected... Ensuring changes have settled before proceeding...
   [[  INF  ]]  2022-03-04 13:47:31 Changes settled... Applying changes and restarting services...
   [[  INF  ]]  Manual certificates have changed, extracting certs..
   [[ TASKS ]]  Setting up SSL
   [[  INF  ]]  TLS configured with 'modern' ciphers
   [[  INF  ]]  Configuring certificates using key /config/ssl/key.ecdsa.pem and cert /config/ssl/cert.ecdsa.pem
   [[  INF  ]]  Configuring fallback certificates using key /config/ssl/key.rsa.pem and cert /config/ssl/cert.rsa.pem
   [[  INF  ]]  SSL configured with 'Manual' certificates
   [[  INF  ]]  Checking file line endings
   [[  INF  ]]  Regenerating postfix user list
   [[  INF  ]]  Creating user 'user1' for domain 'localhost.localdomain'
   [[  INF  ]]  Creating user 'user2' for domain 'otherdomain.tld'
   [[  INF  ]]  Creating user 'user3' for domain 'localhost.localdomain' with attributes 'userdb_mail=mbox:~/mail:INBOX=~/inbox'
   [[  INF  ]]  Adding regexp alias file postfix-regexp.cf
   virtual_alias_maps = texthash:/etc/postfix/virtual pcre:/etc/postfix/regexp
   [[  INF  ]]  Configuring root alias
   [[  INF  ]]  Restarting services due to detected changes..
   sed: 1: "/Users/norsegaud/docker ...": extra characters at the end of n command
```

**Comment by NorseGaud** [comment]
Another two

```
 ✗ File sedfile.bats
   (from function `setup_file' in test file test/sedfile.bats, line 12)
     `sed -i "s|/CONTAINER_START|${CONTAINER_START}|" "${SEDFILE}"' failed
   sed: 1: "/tmp/sedfile": unterminated substitute in regular expression
```

```
 ✗ File tests.bats
   (from function `repeat_until_success_or_timeout' in file test/test_helper/common.bash, line 37,
    from function `repeat_in_container_until_success_or_timeout' in file test/test_helper/common.bash, line 70,
    from function `wait_for_empty_mail_queue_in_container' in file test/test_helper/common.bash, line 164,
    from function `setup_file' in test file test/tests.bats, line 79)
     `wait_for_empty_mail_queue_in_container mail' failed
   7591bfc7f1877ca14ee40cecd7bdc97b7d14f13d218bdfb6349e139b8728d69c
   [ TASKLOG ]  mail.my-domain.com is up and running
   Waiting for dovecot to create /var/mail/localhost.localdomain/pass...
   Waiting for dovecot to create /var/mail/localhost.localdomain/pass...
   Waiting for dovecot to create /var/mail/localhost.localdomain/pass...
   Waiting for dovecot to create /var/mail/localhost.localdomain/pass...
   Waiting for dovecot to create /var/mail/localhost.localdomain/pass...
   Waiting for dovecot to create /var/mail/localhost.localdomain/pass...
   Waiting for dovecot to create /var/mail/localhost.localdomain/pass...
   Waiting for dovecot to create /var/mail/localhost.localdomain/pass...
   Waiting for dovecot to create /var/mail/localhost.localdomain/pass...
   Waiting for dovecot to create /var/mail/localhost.localdomain/pass...
   220 mail.my-domain.com ESMTP
   250 mail.my-domain.com
   250 2.1.0 Ok
   250 2.1.5 Ok
   354 End data with <CR><LF>.<CR><LF>
   250 2.0.0 Ok: queued as 5A39E2062CC8E6
   221 2.0.0 Bye
   220 mail.my-domain.com ESMTP
   250 mail.my-domain.com
   250 2.1.0 Ok
   250 2.1.5 Ok
   354 End data with <CR><LF>.<CR><LF>
   250 2.0.0 Ok: queued as D78092062CC8ED
   221 2.0.0 Bye
   220 mail.my-domain.com ESMTP
   250 mail.my-domain.com
   250 2.1.0 Ok
   250 2.1.5 Ok
   354 End data with <CR><LF>.<CR><LF>
   250 2.0.0 Ok: queued as 761EE2062CC900
   221 2.0.0 Bye
   220 mail.my-domain.com ESMTP
   250 mail.my-domain.com
   250 2.1.0 Ok
   250 2.1.5 Ok
   354 End data with <CR><LF>.<CR><LF>
   250 2.0.0 Ok: queued as 2B89A2062CC925
   221 2.0.0 Bye
   220 mail.my-domain.com ESMTP
   250 mail.my-domain.com
   250 2.1.0 Ok
   250 2.1.5 Ok
   354 End data with <CR><LF>.<CR><LF>
   250 2.0.0 Ok: queued as BCEA22062CC92A
   221 2.0.0 Bye
   220 mail.my-domain.com ESMTP
   250 mail.my-domain.com
   250 2.1.0 Ok
   250 2.1.5 Ok
   354 End data with <CR><LF>.<CR><LF>
   250 2.0.0 Ok: queued as 5EF1D2062CC936
   221 2.0.0 Bye
   220 mail.my-domain.com ESMTP
   250 mail.my-domain.com
   250 2.1.0 Ok
   250 2.1.5 Ok
   354 End data with <CR><LF>.<CR><LF>
   250 2.0.0 Ok: queued as 198EB2062CC93F
   221 2.0.0 Bye
   220 mail.my-domain.com ESMTP
   250 mail.my-domain.com
   250 2.1.0 Ok
   250 2.1.5 Ok
   354 End data with <CR><LF>.<CR><LF>
   250 2.0.0 Ok: queued as ADBC62062CC948
   221 2.0.0 Bye
   220 mail.my-domain.com ESMTP
   250 mail.my-domain.com
   250 2.1.0 Ok
   250 2.1.5 Ok
   354 End data with <CR><LF>.<CR><LF>
   250 2.0.0 Ok: queued as 488222062CC94D
   221 2.0.0 Bye
   220 mail.my-domain.com ESMTP
   250 mail.my-domain.com
   250 2.1.0 Ok
   250 2.1.5 Ok
   354 End data with <CR><LF>.<CR><LF>
   250 2.0.0 Ok: queued as E5C882062CC954
   221 2.0.0 Bye
   220 mail.my-domain.com ESMTP
   250 mail.my-domain.com
   250 2.1.0 Ok
   250 2.1.5 Ok
   354 End data with <CR><LF>.<CR><LF>
   250 2.0.0 Ok: queued as 7D54A2062CC959
   221 2.0.0 Bye
   220 mail.my-domain.com ESMTP
   250 mail.my-domain.com
   250 2.1.0 Ok
   250 2.1.5 Ok
   354 End data with <CR><LF>.<CR><LF>
   250 2.0.0 Ok: queued as 321A92062CC95F
   221 2.0.0 Bye
   220 mail.my-domain.com ESMTP
   250 mail.my-domain.com
   250 2.1.0 Ok
   250 2.1.5 Ok
   354 End data with <CR><LF>.<CR><LF>
   250 2.0.0 Ok: queued as D6D012062CC967
   221 2.0.0 Bye
   220 mail.my-domain.com ESMTP
   250 mail.my-domain.com
   250 2.1.0 Ok
   250 2.1.5 Ok
   354 End data with <CR><LF>.<CR><LF>
   250 2.0.0 Ok: queued as 6951C2062CC96C
   221 2.0.0 Bye
   220 mail.my-domain.com ESMTP
   250 mail.my-domain.com
   250 2.1.0 Ok
   250 2.1.5 Ok
   354 End data with <CR><LF>.<CR><LF>
   250 2.0.0 Ok: queued as EDB8D2062CC972
   221 2.0.0 Bye
   220 mail.my-domain.com ESMTP
   250 mail.my-domain.com
   250 2.1.0 Ok
   550 5.1.1 <nouser@localhost.localdomain>: Recipient address rejected: User unknown in virtual mailbox table
   554 5.5.1 Error: no valid recipients
   221 2.7.0 Error: I can break rules, too. Goodbye.
   Timed out on command: docker exec mail bash -c [[ $(mailq) == *"Mail queue is empty"* ]]
   mail
```

**Comment by casperklein** [comment]
I see, you add more and more Mac specific workarounds. This will be a cat & mouse game forever. Every PR is only tested against linux. So it's just a matter of time, until something breaks on MacOS again (unnoticed, until you stumble upon it.).
Wouldn't it be much easier, to just spin up a dev (linux) virtual machine and use that? That's what I do, instead of adding Windows specific fixes 😆

**Comment by NorseGaud** [comment]
@docker-mailserver/maintainers , since there are a bunch of macOS specific tweaks needed

> I see, you add more and more Mac specific workarounds. This will be a cat & mouse game forever. Every PR is only tested against linux. So it's just a matter of time, until something breaks on MacOS again (unnoticed, until you stumble upon it.). Wouldn't it be much easier, to just spin up a dev (linux) virtual machine and use that? That's what I do, instead of adding Windows specific fixes 😆

I agree. We should pull all custom macOS changes from the code and update the Contributor guide to explain macOS users need to use a Linux VM (I can do in a different PR). What does @georglauterbach and @polarathene think?

**Comment by georglauterbach** [comment]
> > I see, you add more and more Mac specific workarounds. This will be a cat & mouse game forever. Every PR is only tested against linux. So it's just a matter of time, until something breaks on MacOS again (unnoticed, until you stumble upon it.). Wouldn't it be much easier, to just spin up a dev (linux) virtual machine and use that? That's what I do, instead of adding Windows specific fixes 😆
> 
> 
> 
> I agree. We should pull all custom macOS changes from the code and update the Contributor guide to explain macOS users need to use a Linux VM (I can do in a different PR). What does @georglauterbach and @polarathene think?

I very much agree that this should be the way to go :)

**Comment by polarathene** [comment]
> What does @georglauterbach and @polarathene think?

I agree as well. I personally just spin up a VM to run tests.

**Comment by NorseGaud** [comment]
Not stale! I just moved so haven't had time for this

**Comment by georglauterbach** [comment]
> Not stale! I just moved so haven't had time for this

If you want to you can add the `stalebot/ignore` label so GH actions won't mark this PR as stale.

**Comment by georglauterbach** [comment]
This PR is now open for quite some time without recent contribution, and we have quite a lot of merge conflicts here.

I‘d advocate for closing this and, if someone want‘s to try again in the future, open a new PR.

Any objections from @docker-mailserver/maintainers?
